Added Compatibility tests for ExportSpec struct

itay-nakash opened the pull request:
Added different Compatibility tests for the ExportSpec struct

Changed the return value of IsCompatibleWith in exportspec.go to an error-struct
type contains the field that conflicts and the other cluster's id.

itay-nakash committed:
Added Compatibility tests for ExportSpec struct

Added different Compatibility tests for the ExportSpec struct
Changed the return value of IsCompatibleWith in exportspec.go to
Error-struct type contains the field that interfired and the other cluster id

Signed-off-by: itay-nakash <Itay.Nakash@ibm.com>

itay-nakash commented on pkg/mcs/exportspec_test.go:
```diff
@@ -78,25 +97,56 @@ func TestCompatibility(t *testing.T) {
+				return &es
+			}},
+
+		"compatible: local properties": {spec: &sample, compatible: true, //#TODO check if that whay Etai means. Does it
```
I wasn't sure if that's the local properties different that you meant,
I'm not sure if it make sense that two services will have different clusterId so I didnt changed it (does it suppose to be services from different cluster or not?).

elevran replied:
Local properties are not part of the service specification. 
For ExportSpecs of two clusters the ClusterID would by definition be different and they should not conflict.
However - we could (and probably should) change `IsConflictingWith()` to confirm that the namespace and name are identical as well, even though they're local properties. The exports are supposed to represent the *same* service between multiple clusters, so namespace and name ought to match.

elevran replied:
If we add name and namespace to the conflicting fields definition, you should change the test below and also add tests for incompatible name and/or namespace

itay-nakash commented on pkg/mcs/exportspec_test.go:
```diff
@@ -78,25 +97,56 @@ func TestCompatibility(t *testing.T) {
 		other      *mcs.ExportSpec
 		compatible bool
 		field      string
+		mutate     func(mcs.ExportSpec) *mcs.ExportSpec
```
I added the mutate function to edit the samples instead of inserting two samples, as you did in the Marshal tests.
Each time the mutate function changed the sample according to the test needed.

itay-nakash commented on pkg/mcs/exportspec_test.go:
```diff
@@ -78,25 +97,56 @@ func TestCompatibility(t *testing.T) {
-		assertions.Equal(test.field, field, "unexpected field conflict detected")
+		other := test.mutate(*test.spec)
+		errCompatible := test.spec.IsCompatibleWith(other)
+		assertions.Equal(test.compatible, errCompatible == nil) //#TODO make sure its elegant
```
I wasn't sure how you wanted me to use the returned error (some casses do need to return the error, so a simple nil check won't do it). I'm not sure if its an elegant way (or clear way) to write it, but I ended up doing it like this

elevran replied:
Instead of testing compatible and `err == nil`, we can start with `assertions.IsError/IsNil/Error/NotNil` (I don't remember the exact name) and check the return value:
```go
   if test.compatible {
      assertions.NoError(errCompatible)
   else {
      assertions.Error(errCompatible)
   }
```

The exact field name does not matter, but you can add an assertion that the error String is not empty when there's an error. The exact details of the conflict are not interesting from a caller point of view since it'll only set them on the Conflict condition of the `ServiceExport.Status`

elevran replied:
also, just call `errCompatible`  - `err`. No need to be elaborate as it is the only error in the test and you check it immediately after calling.

itay-nakash replied:
> The exact field name does not matter, but you can add an assertion that the error String is not empty when there's an error. The exact details of the conflict are not interesting from a caller point of view since it'll only set them on the Conflict condition of the `ServiceExport.Status`

Shouldn't we make sure that the field that we match as error is the right one (and not accidentally raised error on a different field)?

elevran replied:
I'm debating this with myself... 
From caller perspective it does not really matter, but I agree that detecting the right field should be validated.
How about defining (string) constants for the fields (e.g., NameField, PortField, ...) that can conflict, adding a `Cause()` or `Field()` to the CompatibilityError class so we can compare against that?

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -40,6 +41,16 @@ type GlobalProperties struct {
 	Ports []mcsv1a1.ServicePort `json:"ports"`
 }
 
+type CompatibleError struct {
```
`CompatibilityError`

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -40,6 +41,16 @@ type GlobalProperties struct {
 }
 
+type CompatibleError struct {
+	OtherServiceCluster string
```
these can be made private. We don't need to expose it to callers as directly accessible properties.

elevran replied:
Also, try to use shorter names. especially if private fields. `clusterID` or `cluster` would suffice here.

itay-nakash replied:
Edited it as you asked. 
I used the field as exposed in the tests to compare the field that had error to the field the "actual" field that supposed to had error (But anyway I could have written a method that does it..).

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -40,6 +41,16 @@ type GlobalProperties struct {
 
+type CompatibleError struct {
+	OtherServiceCluster string
+	InterferingField    string
```
`field` is enough since it is in the context of `CompatibilityError`

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -40,6 +41,16 @@ type GlobalProperties struct {
+	InterferingField    string
+}
+
+func (err *CompatibleError) Error() string {
```
please
- drop the `Error` (it is an error by definition...) 
- start with a lower case letter
- remove quotes (or replace with tick (i.e., `'`) if you think them necessary 
- perhaps `field %s conflicts with value in cluster %s`?

elevran commented on pkg/mcs/exportspec_test.go:
```diff
@@ -33,6 +33,25 @@ var (
 		},
 	}
+
+	different_name_sample = mcs.ExportSpec{
```
Is this duplicating the global `sample` variable? Can you reuse it?

itay-nakash replied:
Yes, its useless, I think its trace of the attempts before adding the mutate function (I deleted it now)

itay-nakash committed:
Update the Compatibility tests according to PR comments.

Add additonal tests scenarios
Improve varibles names and prints

Signed-off-by: itay-nakash <Itay.Nakash@ibm.com>

itay-nakash commented on pkg/mcs/exportspec_test.go:
```diff
@@ -136,25 +137,83 @@ func TestCompatibility(t *testing.T) {
+				return &es
+			}},
+
+		"conflicting: affinity config": {spec: &sample, compatible: false, field: "affinity config",
```
I added a test for conflicting affinity config that was missing.

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -139,13 +164,20 @@ func (es *ExportSpec) IsPrefferredOver(another *ExportSpec) bool {
+		return &CompatibilityError{another.ClusterID, AffinityConfigField}
+	}
+
+	if es.Name != another.Name {
```
Pull these (ns and name) up as the first checks, they are more coarse grained (i.e., not the same export so no need to compare field by field)

elevran commented on pkg/mcs/exportspec_test.go:
```diff
@@ -136,25 +137,83 @@ func TestCompatibility(t *testing.T) {
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				es.Service.SessionAffinityConfig = &corev1.SessionAffinityConfig{
+					ClientIP: &corev1.ClientIPConfig{
+						TimeoutSeconds: &otherTimeout,
```
no need to make `otherTimeout` global. It should be a local variable in this (unnamed) function

elevran reviewed (changes requested):
Please make these two small changes and then we can merge.
You may need to rebase again before making the changes since I added a copyright/license at the top of the files

itay-nakash committed:
Update Compatibility tests according to PR comments

Change Compatibility test order in exportspec.go
Made otherTimeout local varible

Signed-off-by: itay-nakash <Itay.Nakash@ibm.com>

elevran commented:
@itay-nakash if you're done with the requested changes, please resolve the relevant conversations and re-request review

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -60,8 +61,32 @@ type GlobalProperties struct {
+		return fmt.Sprintf("field %s conflicts with value in cluster %s",
+			err.field, err.clusterID)
+	}
+	return fmt.Sprintf("The same service has different %s in cluster %s",
```
this is reached when the export specification refers to a different service (name.namespace)?
If so, the error message is confusing (e.g., `"the export refers to a different service and is incompatible"`).

elevran commented on pkg/mcs/exportspec.go:
```diff
@@ -156,13 +181,20 @@ func (es *ExportSpec) IsPreferredOver(another *ExportSpec) bool {
-func (es *ExportSpec) IsCompatibleWith(another *ExportSpec) (bool, string) {
+func (es *ExportSpec) IsCompatibleWith(another *ExportSpec) *CompatibilityError {
+	if es.Name != another.Name {
+		return &CompatibilityError{another.ClusterID, NameField}
```
Please initialize struct fields by name and not implicitly (e.g., we have two strings in this case, what if you accidently reverse them? It won't be caught by the compiler... See https://github.com/uber-go/guide/blob/master/style.md#use-field-names-to-initialize-structs)

itay-nakash committed:
Fix PR comments

Added explicit struct initialization

Changed error message

Signed-off-by: itay-nakash <Itay.Nakash@ibm.com>

elevran reviewed (approved):
lgtm
